File: server/consumer/service/document.py
from sqlalchemy.orm import declarative_base, se
from sqlalchemy import Column, Integer, String, DateTime, select
from .database import Session

Base = declarative_base()

class DownloadError(Exception):
    pass

class DocumentNotFoundError(Exception):
    pass

class ParseError(Exception):
    pass

class AckError(Exception):
    pass

class MessageFormatError(Exception):
    pass

class Document(Base):

    __tablename__ = "documents"

    id = Column(String, primary_key=True)
    created_at = Column(DateTime)
    updated_at = Column(DateTime) 
    deleted_at = Column(DateTime, index=True)
    name = Column(String)
    preview_url = Column(String)
    document_url = Column(String)
    preview_blob_id = Column(String)
    document_blob_id = Column(String)
    size = Column(Integer)
    state = Column(String)
    job_error = Column(String)
    user_id = Column(String)

    def __repr__(self):
        return f"<Document: {self.id}, {self.user_id}, {self.state}, {self.size}"

class DocumentJob:
    __slots__ = ("document_id", "document")
    def __init__(self, document_id):
        self.document_id = document_id
        self.document = None
        self.get_document()
    
    def get_document(self):
        with Session() as session:
            stmt = select(Document).where(Document.user_id == self.document_id)
            for document in session.execute(stmt):
                self.document = document
        if not self.document:
            raise DocumentNotFoundError()
        

File: server/gin/service/document.go
package services

import (
	"time"
	"github.com/google/uuid"
	"gorm.io/gorm"
	"encoding/json"
)

type Document struct {
	ID        	string `gorm:"primaryKey;type:varchar(50)" json:"id"`
	CreatedAt 	time.Time `json:"createdAt"`
	UpdatedAt 	time.Time
	DeletedAt 	gorm.DeletedAt `gorm:"index"`
	Name 	  	string `gorm:"type:text" json:"name"`
	PreviewUrl	string `gorm:"type:text" json:"previewUrl"`
	DocumentUrl string `gorm:"type:text" json:"documentUrl"`
	PreviewBlobId string `gorm:"type:text" json:"previewBlobId"`
	DocumentBlobId string `gorm:"type:text" json:"documentBlobId"`
	Size 		int `gorm:"type:int" json:"size"`
	State 		string `gorm:"type:varchar(25)" json:"state"`
	JobError 	string `gorm:"type:text" json:"jobError"`
	UserID		string `json:"userId"`
}

type DocumentCreation struct {
	UserId 		string `json:"userId" binding:"required"`
	Name   		string `json:"name" binding:"required"`
	PreviewUrl 	string `json:"previewUrl" binding:"required"`
	DocumentUrl string `json:"documentUrl" binding:"required"`
	PreviewBlobId string `json:"previewBlobId" binding:"required"`
	DocumentBlobId string `json:"documentBlobId" binding:"required"`
	Size 		int    `json:"size" binding:"required"`
}

type DocumentJob struct {
	UserId string `json:"userId"`
	DocumentId string `json:"documentId"`
}

const (
	QUEUED = "QUEUED"
)

func GetDocuments(userId string) (*[]Document, error) {
	var documents []Document 
	result := PSQL.DB.Where("user_id = ?", userId).Find(&documents).Order("created_at DESC")
	if result.Error != nil { return nil, result.Error }
	return &documents, nil
}

func CreateDocument(creation DocumentCreation) (*Document, error) {
	documentId := uuid.New().String()
	document := &Document{
		ID: documentId,
		Name: creation.Name,
		PreviewUrl: creation.PreviewUrl,
		DocumentUrl: creation.DocumentUrl,
		PreviewBlobId: creation.PreviewBlobId,
		DocumentBlobId: creation.DocumentBlobId,
		Size: creation.Size,
		State: QUEUED,
		JobError: "",
		UserID: creation.UserId,
	}
	
	if result := PSQL.DB.Create(document); result.Error != nil {
		return nil, result.Error
	} 

	return document, nil
}

func PutDocumentInQueue(job *DocumentJob) error {
	if marshal, err := json.Marshal(job); err != nil {
		return err
	} else {
		return MQ.Push(marshal)
	}
}

func DeleteDocument(documentId string) error {
	return PSQL.DB.Where("id = ?", documentId).Delete(&Document{}).Error
}
